### PR TITLE
Feature/pffile

### PR DESCRIPTION
**Pull request by mhupman**
WIP - Implementation for #7

### CODE REVIEW

**Comment by mhupman** [comment]
While trying to add some proper unit tests for the PFFile functionality I ran into some prickly issues that I'm having trouble resolving...

In addition to being uploaded to S3, the contents of a PFFile are transparently cached on disk when the Parse SDK's `save` method (and its ilk) is invoked.  That could be an issue in itself since it might cause us to not properly test the async behavior of the network request.  However, a more pressing problem is that subsequent calls to `getData` and friends are raising the following exception:

```
/Users/mhupman/repo/Parse-RACExtensions/Parse-RACExtensionsTests/PFFileRACSpec.m:0: error: -[PFFileRACSpecSpec downloading_should_download] : NSInternalInconsistencyException: This file is not ready to be fetched.
  Call Stack:
    0   CoreFoundation                      0x00b1d5e4 __exceptionPreprocess + 180
    1   libobjc.A.dylib                     0x007958b6 objc_exception_throw + 44
    2   CoreFoundation                      0x00b1d3bb +[NSException raise:format:] + 139
    3   Parse-RACExtensionsTests            0x060d529b -[PFFile assertAwaitingFetchOrFetching] + 83
    4   Parse-RACExtensionsTests            0x060d668d -[PFFile getDataAsync:] + 35
    5   Parse-RACExtensionsTests            0x060d6616 -[PFFile getDataInBackgroundWithBlock:progressBlock:] + 41
    ...snip...
```

My guess is that since I'm attempting to `fetch` a file that was just saved from the same device, it fails the `[PFFile assertAwaitingFetchOrFetching]` check inside the Parse SDK.  Inspecting  `isDataAvailable` and `isDirty` support the assumption that the Parse SDK thinks the file is fully downloaded and doesn't need fetching.

A solution to this problem would be to find a way to obtain a handle to a PFFile that represents a valid remote resource but has not already been downloaded and cached locally.  But...
- PFFile doesn't expose a way to create a valid PFFile pointer (w/o data) from an existing PFFile instance.  Something like `[PFFile fileWithIdentifer:(NSString*)guid]` would be ideal.
- The Parse SDK doesn't provide a mechanism to clear cached file data.  You can find the on-disk cache via `NSCachesDirectory`, but deleting it has no effect on `isDataAvailable`.
- Specific to the unit tests, I can't think of any way to "clean up" the Parse account after the test, like the `PFObjectRACSpec` does.  PFFile doesn't have a `delete` method, and Parse documentation says you cannot delete a PFFile that isn't attached to a PFObject ( https://parse.com/docs/ios_guide#files-progress/iOS).

Thoughts?

**Comment by kastiglione** [comment]
@mhupman – I sent you a message out of band, did you get it?

**Comment by mhupman** [comment]
I did not - where did you send it?  I went ahead and added my email address to my public profile.

**Comment by kastiglione** [comment]
Cool, I'll send you an email right now.

**Comment by a2** [comment]
What's the state of this PR? Coming soon?

**Comment by mhupman** [comment]
@a2 I meant to get someone to review but it fell off of my radar.  If you, @jondwillis or @MattCBowman want to give it a once over that would be appreciated.  I believe its working as intended, but the unit-tests are prickly (as you can see from the thread above) and I haven't had the chance to use it in a real-world project.

**Comment by kastiglione** [comment]
oh hai

EDIT: I'll get involved in this asap, I'll give an update by Monday or Tuesday :sparkling_heart:

**Comment by kastiglione** [comment]
@mhupman Can you temporarily hold off on merging this. I still hope to have more details next week.

**Comment by mhupman** [comment]
Sure thing!

**Comment by kastiglione** [comment]
Still no update, but I haven't forgotten and will see if I can get an update soon.

**Comment by kastiglione** [comment]
@mhupman Things are progressing now, and I'm cautiously optimistic I can review and get the ball really rolling the end of next week. If not then the week after. I haven't looked too closely, but the testing approach looks like a pretty good solution.

File: Parse-RACExtensions/PFFile+RACExtensions.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+//
+//  Created by Matt Hupman on 2014-02-05.
+//  Copyright (c) 2014 Matt Hupman. All rights reserved.
+//
+
+#import <Parse/PFFile.h>
+
+@class RACSignal;
+
+@interface PFFileDownload : NSObject
```
Comment (kastiglione) [inline Parse-RACExtensions/PFFile+RACExtensions.h:13]: Did you look into `NSProgress` for this purpose?

Comment (mhupman) [inline Parse-RACExtensions/PFFile+RACExtensions.h:13]: I did not, _mostly_ due the fact that I wasn't aware of its existence.  I'll take a look!

Comment (mhupman) [inline Parse-RACExtensions/PFFile+RACExtensions.h:13]: We currently support back to [iOS 5.0](https://github.com/kastiglione/Parse-RACExtensions/blob/master/Parse-RACExtensions.podspec#L9), but `NSProgress` is 7.0+ only - is it worth bumping the target?

Comment (kastiglione) [inline Parse-RACExtensions/PFFile+RACExtensions.h:13]: That support for iOS 5 is to follow that which RAC supports.

RAC 3 is iOS 6+. RAC 2 might also move to 6+.

But I'd be open to move to iOS 7+. If a user came along and said they wanted to use the library but needed iOS 6, it seems quite possible that `NSProgress` could be back-ported to iOS 6, similar to what `PSTCollectionView` to allow `UICollectionView` on iOS 5.

**Comment by kastiglione** [comment]
@mhupman Thanks for your contributions. Given the news that Parse is being shut down, I'm going to close this.

**Comment by gfosco** [comment]
The reports of our death have been greatly exaggerated.    https://github.com/ParsePlatform/parse-server

**Comment by kastiglione** [comment]
@gfosco Thanks! Apologies for making it appear that I was closing this solely because of Parse transitioning away from a commercial support.

This project hasn't been in active development for some time now, and probably not in active use either. The Parse SDKs have seen many updates, and ReactiveCocoa is now on version 4 and uses Swift. All of these combined led me to the conclusion that it would be best to officially stop this project.

**Pull request closed without merging**